scalding-core/src/main/scala/com/twitter/scalding/typed/memory_backend/MemoryBackend.scala

```diff
@@ -0,0 +1,527 @@
+   * fn may be called more than once
+   */
+  def update[R](fn: T => (T, R)): R = {
+
+    @annotation.tailrec
+    def loop(): R = {
+      val init = ref.get
+      val (next, res) = fn(init)
+      if (ref.compareAndSet(init, next)) res
+      else loop()
+    }
+
+    loop()
+  }
+
+  def get(): T = ref.get
+}
+
+final class MemoryMode private (srcs: Map[TypedSource[_], Iterable[_]], sinks: Map[TypedSink[_], AtomicBox[Option[Iterable[_]]]]) extends Mode {
+
+  def newWriter(): Writer =
+    new MemoryWriter(this)
+
+  def openForRead(config: Config, tap: Tap[_, _, _]): TupleEntryIterator = ???
+  def fileExists(filename: String): Boolean = ???
+  def newFlowConnector(props: Config): FlowConnector = ???
+
+  def addSource[T](src: TypedSource[T], ts: Iterable[T]): MemoryMode =
+    new MemoryMode(srcs + (src -> ts), sinks)
+
+  def addSink[T](sink: TypedSink[T]): MemoryMode =
+    new MemoryMode(srcs, sinks + (sink -> new AtomicBox[Option[Iterable[_]]](None)))
+
+  /**
+   * This has a side effect of mutating this MemoryMode
+   */
+  def writeSink[T](t: TypedSink[T], iter: Iterable[T]): Unit =
+    sinks(t).lazySet(Some(iter))
+
+  def readSink[T](t: TypedSink[T]): Option[Iterable[T]] =
+    sinks.get(t).flatMap(_.get).asInstanceOf[Option[Iterable[T]]]
+
+  def readSource[T](t: TypedSource[T]): Option[Iterable[T]] =
+    srcs.get(t).asInstanceOf[Option[Iterable[T]]]
+}
+
+object MemoryMode {
+  def empty: MemoryMode = new MemoryMode(Map.empty, Map.empty)
+}
+
+object MemoryPlanner {
+
+  sealed trait Op[+O] {
+    def result(implicit cec: ConcurrentExecutionContext): Future[ArrayBuffer[_ <: O]]
+
+    def flatMap[O1](fn: O => TraversableOnce[O1]): Op[O1] =
+      Op.Map(this, fn)
+
+    def mapAll[O1 >: O, O2](fn: IndexedSeq[O1] => ArrayBuffer[O2]): Op[O2] =
+      Op.MapAll[O1, O2](this, fn)
+  }
+  object Op {
+    def source[I](i: Iterable[I]): Op[I] = Source(_ => Future.successful(i))
+    def empty[I]: Op[I] = source(Nil)
+
+    case class Source[I](input: ConcurrentExecutionContext => Future[Iterable[I]]) extends Op[I] {
+      private[this] val promise: Promise[ArrayBuffer[I]] = Promise()
+
+      def result(implicit cec: ConcurrentExecutionContext): Future[ArrayBuffer[I]] = {
+        if (!promise.isCompleted) {
+          promise.tryCompleteWith {
+            val iter = input(cec)
+            iter.map { i => ArrayBuffer.concat(i) }
+          }
+        }
+
+        promise.future
+      }
+    }
+
+    case class Materialize[O](op: Op[O]) extends Op[O] {
+      private[this] val promise: Promise[ArrayBuffer[_ <: O]] = Promise()
+
+      def result(implicit cec: ConcurrentExecutionContext) = {
+        if (!promise.isCompleted) {
+          promise.tryCompleteWith(op.result)
+        }
+
+        promise.future
+      }
+    }
+
+    case class Concat[O](left: Op[O], right: Op[O]) extends Op[O] {
+      def result(implicit cec: ConcurrentExecutionContext) = {
+        // start both futures in parallel
```

Comment:
bad comment from when I did `for { }` which is not parallel unless we start the futures first. Now, it doesn't matter (using zip).